Objects/floatobject.c

```diff
@@ -1208,11 +1201,116 @@ hex_from_char(char c) {
     return x;
 }
 
-/* convert a float to a hexadecimal string */
+/* Convert a float to a hexadecimal string [±][0x]h[.hhhhhhhh]p±d,
+   where the fractional part either is exact (precision < 0) or the
+   number of digits after the dot is equal to the precision.
+
+   The exponent d is written in decimal, it always contains at least one digit,
+   and it gives the power of 2 by which to multiply the coefficient.
+
+   x - the double to be converted
+   precision - the desired precision
+   always_add_sign - nonzero if a '+' sign should be included for x > 0
+   use_alt_formatting - nonzero if the hexadecimal prefix should be added.
+   upper - nonzero, if uppercase letters should be used for hexadecimal
+           numbers, prefix and the exponent separator.
+   _hex - private flag to keep float.hex() behaviour
+ */
+
+char *
+_Py_dg_dtoa_hex(double x, int precision, int always_add_sign,
+                int use_alt_formatting, int upper, int _hex)
+{
+    int e, autoprec = precision < 0;
+    double m = frexp(fabs(x), &e);
+
+    if (autoprec) {
+        precision = (DBL_MANT_DIG + 2 - (DBL_MANT_DIG+2)%4)/4;
+        if (!x && _hex) {
+            /* for compatibility with float.hex(), we keep just one
+               digit of zero */
+            precision = 1;
+        }
+    }
+
+    /* normalization */
+    if (m) {
+        int shift = 1 - Py_MAX(DBL_MIN_EXP - e, 0);
+        m = ldexp(m, shift);
+        e -= shift;
+    }
+
+    /* round to precision digits */
+    if (!autoprec) {
+        do {
+            double frac = ldexp(m, 4*precision);
+            frac -= floor(frac);
+            frac *= 16.0;
+            m += ldexp(frac >= 8.0, -4*precision);
```

Comment:
`frac >= 8.0` result is 1 or 0? Maybe write it more explicitly?

```suggestion
            if (frac >= 8.0) {
                m += ldexp(1.0, -4*precision);
            }
```